Rossby Test Case Initialization

jjuyeonkim opened the pull request:
**Description**
This is the first step in porting a Fortran shallow water test to pace. This PR will add DycoreState initialization for the Rossby-Haurwitz wave 4 test case found in `tools/test_cases.F90` (shallow-water test 6) of the [GFDL_atmos_cubed_sphere repo](https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere.git).

Additionally, there is an new `sw_dynamics` configuration flag added to the DynamicalCoreConfig that does nothing yet, but it will eventually be used to modify logic based on whether or not the test is for shallow water.

Fixes # (issue) 
None.

**How Has This Been Tested?**
DycoreState initialization for the Rossby case was unit-tested in the pace repository. This related code will be added in a separate PR for that repo.

**Checklist:**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation - _I don't believe there is corresponding documentation for this change._
- [x] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules - _Dependent changes will be merged eventually, but haven't been merged yet._
- [ ] New check tests, if applicable, are included - _Unit test will be added in a separate repository PR_
- [ ] Targeted model if this changed was triggered by a model need/shortcoming - _Unsure. I don't think this is applicable._

jjuyeonkim committed:
Rossby initialization

jjuyeonkim committed:
Linting Rossby initialization

jjuyeonkim committed:
Adding sw_dynamics to DynamicalCoreConfig

jjuyeonkim committed:
Merge branch 'develop' into 20250421_rossby_2

jjuyeonkim committed:
Linting rossby init

jjuyeonkim commented:
This PR is partially tested in a related PR: https://github.com/NOAA-GFDL/pace/pull/112

FlorianDeconinck commented on pyFV3/initialization/test_cases/initialize_rossby.py:
```diff
@@ -0,0 +1,191 @@
+Corresponds to Fortran shallow-water test #6 found in tools/test_cases.F90 of
+https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere.git
+"""
+
```
This will always generate a 64-bit float state. Not knowing the science I don't know if it's an issue but I flag it nonetheless

jjuyeonkim replied:
Thanks for this! This is a case that I don't understand well. 
I'm not sure how to best handle float32 vs float64 precision in the pace code. (@fmalatino Frank did show me the 'Float' in NDSL's typing.py, so thanks for that!) @oelbert -- I was wondering if I should add another unit test that forces and checks float32 precision? Or, if there's something that's usually done, please let me know! Thanks everyone.

FlorianDeconinck replied:
As you know, Python doesn't have a 32-bit float. So we are tied to using the `Float` caster which basically uses C (wrapped by numpy) to carry proper 32-bit float.

The rule of thumb is for Fortran port is: if the `real` type is not specifying a preicsion (e.g. `real(kind=X)`) then it's a compiler defined `real`, and it maps to `Float` in our system. If the variable is strongly type `real(kind=X)` OR the initialize value is typed (e.g. `3.14d`) then you have to strongly type it on the pythong using `f32` or `f64` so the proper type carries and it is made verbose that those value have an inherent precision

oelbert replied:
The baroclinic instability test case also generates states at 64-bit precision; should we update that as well (probably not in this PR)?

oelbert replied:
Maybe it makes sense to open an issue and switch all of the test cases over at once when @lharris4  and @bensonr have weighed in?

FlorianDeconinck replied:
We will probably need a way to cast-down any initialization to the runtime precision, if we keep calculation in pure `numpy/cupy`.

Another solution is to move calculation under stencils/ndsl, and then precision will follow. But this will require scientific validation.

oelbert replied:
actually I take it back, `init_utils.empty_numpy_dycore_state` initializes a dycore state with `Float` precision already --shouldn't that handle everything?

lharris4 replied:
I would prefer to be able to run the test cases with both single and double runtime precision. Luan is currently finding some interesting behavior with some single-precision tests.

bensonr replied:
Completely agree with @lharris4.  Being able to run the tests in both precisions to test for _"correctness"_ and examine the conservation properties is essential.

FlorianDeconinck replied:
The question that remains is: do we need the _internals_ of the test case numerics to be computed at 32-bit? Or we just need the _results_ of initialization (the state) to be at the right precision.

The first one will require either moving to NDSL or type casting floats. The second is already there.

bensonr replied:
I'll let @lharris4 answer definitively, but I'd want the ICs to create a state that allows the numerics of the test to proceed at the desired precision.

FlorianDeconinck replied:
And that should already be the case, as the DycoreState enforce precision, even if the incoming data is 64-bit (it'll be cast on copy)

lharris4 replied:
@bensonr @FlorianDeconinck I agree.

oelbert replied:
I think this means we're good to go here?

jjuyeonkim replied:
After FV3/MSD meeting, we've decided to do the internal calculation at compiler precision (using PACE_FLOAT_PRECISION). This will required some rework of the rossby test to cast to NDSL Float. We will open an issue to make sure these changes are applied to other tests (Baroclinic, TropicalCyclone) -- https://github.com/NOAA-GFDL/PyFV3/issues/53

FlorianDeconinck commented on pyFV3/initialization/analytic_init.py:
```diff
@@ -60,6 +61,16 @@ def init_analytic_state(
+        elif analytic_init_case == Cases.rossby.value:  # type: ignore
+            import pyFV3.initialization.test_cases.initialize_rossby as rossby
+
+            assert isinstance(comm, CubedSphereCommunicator)
```
Turn into a try/catch with a `ValueError` describing that `rossby` is only available on the cube sphere

jjuyeonkim replied:
For this one, I replaced the assert to raise a TypeError. I hope this is okay.

FlorianDeconinck commented on pyFV3/_config.py:
```diff
@@ -274,6 +274,7 @@ class DynamicalCoreConfig:
     nf_omega: int = NamelistDefaults.nf_omega
     fv_sg_adj: int = NamelistDefaults.fv_sg_adj
     n_sponge: int = NamelistDefaults.n_sponge
+    sw_dynamics: bool = DEFAULT_BOOL
```
Do we want to `NameslistDefaults` this? I am not a fan of the process tbh, so a stern no would sway me easily :)

jjuyeonkim replied:
I think your suggestion makes sense. I'll modify this to use NamelistDefaults.sw_dynamics. I'll create a PR in NDSL soon. Thanks!

FlorianDeconinck reviewed (changes requested):
Small assert->try/catch but we are looking good.

jjuyeonkim committed:
Mods based on https://github.com/NOAA-GFDL/PyFV3/pull/50
- Use NamelistDefaults.sw_dynamics
- Replace assert with raised TypeError in analytic_init.py

oelbert commented on pyFV3/initialization/analytic_init.py:
```diff
@@ -35,11 +39,13 @@ def init_analytic_state(
-            import pyFV3.initialization.test_cases.initialize_baroclinic as bc
-
-            assert isinstance(comm, CubedSphereCommunicator)
+        if not isinstance(comm, CubedSphereCommunicator):
+            raise TypeError(
+                f"Expected CubedSphereCommunicator instance for 'comm', "
+                f"got {type(comm).__name__} instead."
+            )
```
There will eventually be some test cases that use a doubly-periodic TileCommunicator. We could have a check on whether the test case is in a group of spherical or single-tile cases here instead of checking within each case branch if you like?

jjuyeonkim committed:
Grouping spherical analytic test Cases

jjuyeonkim committed:
Using NDSL Float for correct precision

jjuyeonkim committed:
Rossby Init brute force u/v slicing to avoid Nans

jjuyeonkim committed:
Cleaning up delp, u+v slicing

jjuyeonkim committed:
Merge branch 'develop' into 20250421_rossby_2

oelbert reviewed (dismissed):
This looks good to me. Looks like the failing test is just because the NDSL version being pulled doesn't have sw_dynamics but thta's easy to fix

jjuyeonkim committed:
Hard-coding sw_dynamics to False, will change to NamelistDefaults.sw_dynamics with the next NDSL release

jjuyeonkim commented:
> This looks good to me. Looks like the failing test is just because the NDSL version being pulled doesn't have sw_dynamics but thta's easy to fix

For now, I've hard-coded the sw_dynamics flag to False, but I left a TODO to change it back to the NamelistDefaults once there is a new NDSL tag. The small change to add the flag to NDSL is in 'develop' currently.

oelbert commented:
@FlorianDeconinck ready for re-review

FlorianDeconinck reviewed (dismissed):
Looking good (sorry for delay).

One thing you want to add is a try/catch around `sw_dynamics` being `True` when we try to init the `Rossby` case so bad config are caught early.

jjuyeonkim commented:
> One thing you want to add is a try/catch around `sw_dynamics` being `True` when we try to init the `Rossby` case so bad config are caught early.

I think you're right. We should  check that `sw_dynamics` is `True`. Is it possible to push this PR for now. Then, open a separate PR for this check? 

The reason I ask is that I'm currently thinking of modifying the input parameters to `init_analytic_state(...)` to take in a config. This would likely require a coordinated change in pace. I was hoping to create a new PR for this part.

jjuyeonkim committed:
Adding TODO: Check sw_dynamics is True for Rossby

FlorianDeconinck reviewed (approved):
LGTM
